src/Sylius/Bundle/ApiBundle/EventSubscriber/ProductDeletionEventSubscriber.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\EventSubscriber;
+
+use ApiPlatform\Core\EventListener\EventPriorities;
+use Sylius\Bundle\ApiBundle\Exception\ProductCannotBeRemoved;
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Core\Promotion\Checker\ProductPromotionRuleCheckerInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\ViewEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+final class ProductDeletionEventSubscriber implements EventSubscriberInterface
+{
+    public function __construct(private ProductPromotionRuleCheckerInterface $productPromotionRuleChecker)
+    {
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            KernelEvents::VIEW => ['protectFromRemovingProductAssignedToPromotionRule', EventPriorities::PRE_WRITE],
+        ];
+    }
+
+    public function protectFromRemovingProductAssignedToPromotionRule(ViewEvent $event): void
```

Comment:
We could stick to the naming next time, as we use `assigned/used` interchangeably, but that's not a critical suggestion.